testing/mock/_asserts_test.ts

```diff
@@ -0,0 +1,932 @@
+import { AssertionError, assertRejects, assertThrows } from "../asserts.ts";
+import {
+  assertSpyCall,
+  assertSpyCallArg,
+  assertSpyCallArgs,
+  assertSpyCallAsync,
+  assertSpyCalls,
+} from "./_asserts.ts";
+import { Point } from "./_test_utils.ts";
+import { spy, stub } from "./mock.ts";
+
+Deno.test("assertSpyCalls", () => {
+  const spyFunc = spy();
+
+  assertSpyCalls(spyFunc, 0);
+  assertThrows(
+    () => assertSpyCalls(spyFunc, 1),
+    AssertionError,
+    "spy not called as much as expected",
+  );
+
+  spyFunc();
+  assertSpyCalls(spyFunc, 1);
+  assertThrows(
+    () => assertSpyCalls(spyFunc, 0),
+    AssertionError,
+    "spy called more than expected",
+  );
+  assertThrows(
+    () => assertSpyCalls(spyFunc, 2),
+    AssertionError,
+    "spy not called as much as expected",
+  );
+});
+
+Deno.test("assertSpyCall function", () => {
+  const spyFunc = spy((multiplier?: number) => 5 * (multiplier ?? 1));
+
+  assertThrows(
+    () => assertSpyCall(spyFunc, 0),
+    AssertionError,
+    "spy not called as much as expected",
+  );
+
+  spyFunc();
+  assertSpyCall(spyFunc, 0);
```

Comment:
Is this call signature useful? It doesn't seems obvious to me that what is asserted with this call

How about making the 3rd argument non-optional?

Reply:
The third argument is optional. In `assertSpyCall`, the first check it does is that it verifies there were enough calls for the index passed in to exist. By calling it without a third arg or with empty object for the third arg, you are basically asking it to assert it was called enough times for `spyFunc.calls[index]` to exist and it will tell you how many times it was actually called if it doesn't.

Below is an example demonstrating how this is useful over just accessing the calls directly on the `spyFunc`. `assertSpyCall` uses `assertEquals` for it's arguments. If you wanted to do a strict equals comparison on one of the arguments, you would need to get the call and make the assertion on the arg like shown below.

```ts
const call = spyFunc.calls[0];
assertStrictEquals(call.args[0], thing);
```

The problem with this is that if `spyFunc` wasn't called as many times as expected, call will be undefined and it would throw because `args` doesn't exist on undefined. You could add optional chaining to avoid it throwing that way but then it would throw an error where you don't know if it's because `call?.args[0]` was undefined or if the `spyFunc` just wasn't called that many times.

```ts
const call = assertSpyCall(spyFunc, 0);
assertStrictEquals(call.args[0], thing);
```

With `assertSpyCall`, if spyFunc wasn't called enough times it would give you a clear error saying how much it was actually called and if it was called enough times you can confidently know that if the next assertion fails because `call.args[0]` is undefined, you would know the function was called incorrectly.

In the above example, the first line could also have other assertions about the call such as what `this` was or what it returned when it was called.